File: js/128/104.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *   this.val = val;
 *   this.left = this.right = null;
 * }
 */

/**
 * 递归
 * @param {TreeNode} root
 * @param {number} depth
 * @return {number}
 */
const maxDepth = function(root, depth = 0) {
  if(root) {
    depth++
    let ld = depth, rd = depth
    if(root.left) {
      ld = maxDepth(root.left, depth)
    }
    if(root.right) {
      rd = maxDepth(root.right, depth)
    }
    depth = Math.max(ld, rd)
  }
  return depth
}

File: py3/128/104.py
from collections import deque
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # BFS
    def maxDepthBFS(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        que = deque([root])
        depth = 0
        while que:
            n = len(que)
            for _ in range(n):
                node = que.popleft()
                if node.left: que.append(node.left)
                if node.right: que.append(node.right)
            depth += 1

        return depth

    # DFS
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(node, depth):
            if not node: return depth
            depth += 1
            return max(dfs(node.left, depth), dfs(node.right, depth))

        return dfs(root, 0)